ipfs-8-to-9/migration/migration.go

```diff
@@ -0,0 +1,242 @@
+			log.Error(err)
+			return err
+		}
+	} else { // backup file exists
+		log.Log("WARN: backup file %s already exists. CIDs-Multihash pairs will be appended", backupPath)
+	}
+
+	// If it exists, append to it.
+	f, err := os.OpenFile(backupPath, os.O_WRONLY|os.O_APPEND|os.O_CREATE, 0644)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	defer f.Close()
+	buf := bufio.NewWriter(f)
+	defer buf.Flush()
+
+	swapCh := make(chan Swap, 1000)
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for sw := range swapCh {
+			fmt.Fprint(buf, sw.Old.String()+","+sw.New.String()+"\n")
+		}
+	}()
+
+	cidSwapper := CidSwapper{Store: blocks, SwapCh: swapCh}
+	total, err := cidSwapper.Run()
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	close(swapCh)
+	wg.Wait()
+
+	log.Log("%d CIDv1 keys swapped to raw multihashes", total)
+	if err := repo.WriteVersion("9"); err != nil {
+		log.Error("failed to write version file")
+		return err
+	}
+	log.Log("updated version file")
+
+	return nil
+}
+
+// Revert attempts to undo the migration using the log file written by Apply.
+func (m Migration) Revert(opts migrate.Options) error {
+	log.Verbose = opts.Verbose
+	log.Log("reverting %s repo migration", m.Versions())
+
+	lk, err := m.lock(opts)
+	if err != nil {
+		return err
+	}
+	defer lk.Close()
+
+	repo := mfsr.RepoPath(opts.Path)
+
+	log.VLog("  - verifying version is '9'")
+	if err := repo.CheckVersion("9"); err != nil {
+		return err
+	}
+
+	log.VLog("  - starting raw multihash to CIDv1 block migration")
+	dstore, err := m.open(opts)
+	if err != nil {
+		return err
+	}
+	defer dstore.Close()
+	blocks := namespace.Wrap(dstore, ds.NewKey("/blocks"))
+
+	// Open revert path for reading
+	backupPath := filepath.Join(opts.Path, backupFile)
+	log.VLog("  - backup file will be read from %s", backupPath)
+	f, err := os.Open(backupPath)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	unswapCh := make(chan Swap, 1000)
+	scanner := bufio.NewScanner(f)
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		defer close(unswapCh)
+
+		for scanner.Scan() {
+			line := scanner.Text()
+			oldAndNew := strings.Split(line, ",")
+			if len(oldAndNew) != 2 {
+				log.Error("bad line in backup file: %s", line)
+				continue
+			}
+			sw := Swap{Old: ds.NewKey(oldAndNew[0]), New: ds.NewKey(oldAndNew[1])}
+			unswapCh <- sw
+		}
+		if err := scanner.Err(); err != nil {
+			log.Error(err)
+			return
+		}
+
+	}()
+
+	cidSwapper := CidSwapper{Store: blocks}
+	total, err := cidSwapper.Revert(unswapCh)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	wg.Wait()
+
+	log.Log("%d multihashes reverted to CidV1s", total)
+	if err := repo.WriteVersion("8"); err != nil {
+		log.Error("failed to write version file")
+		return err
+	}
+
+	log.Log("reverted version file to version 8")
+	f.Close()
+	err = os.Remove(backupPath)
+	if err != nil {
+		log.Error("could not remove the backup file, but migration worked: %s", err)
+	}
```

Comment:
TODO: Not sure if migration should end up in error if file cannot be removed. Not sure if file should be removed after revert.

Reply:
I agree. If we fail to remove it, leave it. However, if we're trying to go _forward_, we need to make sure to replace/remove this file before we even start.

Reply:
Current version renames it, and returns error if that does not work, but the migration should be ok, and if it is moving forward again, and the user did not care to touch the file, it will just append to it, which is still fine even if it was used for another revert (will just take longer).

I really hope no one runs a migration followed by a revert which actually only fails on the very last step of renaming a file, anyways.

Reply:
I'm leaving this one as is.